### PR TITLE
fix conformal age and distance quantities in bg

### PR DESCRIPTION
**Pull request by nickhand**
Does this look right to you @rainwoodman ? The conformal age of the universe was be computed at a_max and not a_today, and all of the distances were computed using that value and were thus inaccurate. 

I am going to add some tests against astropy (which can do some negative z distance estimates).

### CODE REVIEW

**Comment by nickhand** [comment]
Also @rainwoodman, take a look at bac3fc3. I am a bit confused by the Omega and rho definitions you were using. It seems the values output by the ``index_bg_Omega_m`` and ``index_bg_Omega_r`` aren't what I expected. With the changes in bac3fc3, the values now agree with astropy and the expected scalings I would expect for the Omegas, i.e., (1+z)^n / E(z)^2 with n=3,4 for matter, radiation

File: classylss/binding.pyx [hunk 1]
```diff
@@ -666,62 +684,66 @@ cdef class Background:
         Density parameter of relativistic (radiation like) component, including
         relativistic part of massive neutrino and massless neutrino.
         """
-        return self.compute_for_z(z, self.ba.index_bg_Omega_r)
+        E = self.efunc(z)
```
Comment (rainwoodman) [inline classylss/binding.pyx]: This is wrong. Omega_r contains the pressure part of massive neutrino which is redshift dependent. So the scaling against (1+z) **4 is wrong.

File: classylss/binding.pyx [hunk 2]
```diff
@@ -571,14 +589,14 @@ cdef class Background:
...
-        return self.Omega_m(z) * self.rho_tot(z)
+        return self.Omega_m(z) * self.rho_crit(z)
 
     def rho_r(self, z):
         r"""
         Density of radiation :math:`\rho_r` as a function of redshift, in
         units of :math:`10^{10} (M_\odot/h) (Mpc/h)^{-3}`.
         """
-        return self.Omega_r(z) * self.rho_tot(z)
+        return self.Omega_r(z) * self.rho_crit(z)
```
Comment (rainwoodman) [inline classylss/binding.pyx]: Did you check if this is in class? I think the definition in class is against rho_tot, not rho_crit (I may have miss looked).

File: classylss/binding.pyx [hunk 3]
```diff
@@ -666,62 +684,66 @@ cdef class Background:
...
 
     def Omega_m(self, z):
         """
         Density parameter of non-relativistic (matter like) component, including
         non-relativistic part of massive neutrino. Unit
         """
-        return self.compute_for_z(z, self.ba.index_bg_Omega_m)
+        E = self.efunc(z)
+        z = np.array(z, dtype=np.float64)
+        return self.Omega0_m * (1+z)**3 / E**2
```
Comment (rainwoodman) [inline classylss/binding.pyx]: Same reason as above, matter contains a neutrino part that is redshift dependent, so you can't scale it from today's value.

File: depends/class-2.6.0-a_max.patch [hunk 4]
```diff
@@ -43,10 +43,22 @@ diff -uNr class_public-2.6.0.old/source/background.c class_public-2.6.0/source/b
...
                                 &last_index,
                                 pvecback_integration,
                                 pba->bi_size,
+@@ -1607,8 +1607,9 @@
+              pba->error_message);
+
+   /* substitute last line with quantities today */
++  /* NOTE: comment out b/c last index refers to a_max, not a_today
+   for (i=0; i<pba->bi_size; i++)
+-    pData[(pba->bt_size-1)*pba->bi_size+i]=pvecback_integration[i];
```
Comment (rainwoodman) [inline depends/class-2.6.0-a_max.patch]: The last index on the interplation table must be a_max if you replace it with a_today all hell is wrong.. unless you force a_today == a_max, but that's not allowing going to the future -- in class convention a/a_today is what we usually call a.

Comment (rainwoodman) [reply depends/class-2.6.0-a_max.patch]: `all hell is wrong` == the variable of the table is no longer monotonic a decreases from a_max to a_today near the last item.

**Comment by rainwoodman** [comment]
I think astropy is wrong -- it is base on the assumption that massive neutrino has no pressure like contribution. Because we use class to get efunc_prime anyways, there is no good reason to require Omega_m to scale like a**3.  I would side with the class convention.

**Comment by nickhand** [comment]
Okay, a couple of things, @rainwoodman:

- I've updated the patch to split the background loop up to a_today and then continue up to a_max. It computes the needed a_today values after the first loop and also ensures that the results up to a_max are identical when using a_max=1.0 or a_max>1.0
- You are right that CLASS is using Omega defined with respect to rho_tot internally. I am not sure what the reason is for that, since I thought rho_crit was the accepted definition. I think we should use rho_crit for the binding. Thoughts? 

The tests against are astropy are now all green when using the ``rho_crit`` definition of Omega

**Comment by rainwoodman** [comment]
I like your approach to the the a_today / a_max split. I think we can merge this, but we shall ask the class author why Omega's are relative to rho_tot not rho_crit -- could it be related to the sign of rho_k?

**Comment by rainwoodman** [comment]
My additional worry is that when our same named functions is close but subtly different from class's, a lot of confusions will rise down the road. (this is different froma unit conversion which is usually more obvious to spot).

**Comment by nickhand** [comment]
Yes that is something to worry about I think, although this is all only matters when there is curvature and there is usually not. 

Perhaps we are missing something with the Omega definition? Omega_m and Omega_r are the only ones defined in the code, right? I couldn't find any reference in the code or in their paper. 
Do you want to open up an issue?

**Comment by rainwoodman** [comment]
Here is how rho_tot is computed in background.c
```
  double rho_tot;
  rho_tot = 0.;
  rho_tot += pvecback[pba->index_bg_rho_g];
  rho_tot += pvecback[pba->index_bg_rho_b];
    rho_tot += pvecback[pba->index_bg_rho_cdm];
    rho_tot += pvecback[pba->index_bg_rho_dcdm];
    rho_tot += pvecback[pba->index_bg_rho_dr];
    rho_tot += pvecback[pba->index_bg_rho_scf];
    //printf(" a= %e, Omega_scf = %f, \n ",a_rel, pvecback[pba->index_bg_rho_scf]/rho_tot );
      rho_tot += rho_ncdm;
    rho_tot += pvecback[pba->index_bg_rho_lambda];
    rho_tot += pvecback[pba->index_bg_rho_fld];
    rho_tot += pvecback[pba->index_bg_rho_ur];
  pvecback[pba->index_bg_H] = sqrt(rho_tot-pba->K/a/a);
  pvecback[pba->index_bg_H_prime] = - (3./2.) * (rho_tot + p_tot) * a + pba->K/a;
  pvecback[pba->index_bg_Omega_r] = rho_r / rho_tot;
    pvecback[pba->index_bg_rho_crit] = rho_tot-pba->K/a/a;
    pvecback[pba->index_bg_Omega_m] = rho_m / rho_tot;
```
Omega_r and Omega_m are defined as ratios to rho_tot -- so if we want use these then we need to use rho_tot.

It is only a matter of normalization convention -- whether Omega_tot = 1.0 or Omega_crit = 1.0.

 I do have the impression that CLASS knows more what they are doing than the astropy people in terms of cosmology parameters... I am guessing there might be some advantages setting Omega_crit = 1.0 with DE fluid or dr / scf.

**Comment by nickhand** [comment]
Omega_m and Omega_r both enter into the halo fit calculation in the code, but this assumes zero curvature so this wouldn't be an issue there. The only other place this enters is for Omega_r in a calculation of Neff here:

https://github.com/lesgourg/class_public/blob/5908e1c3d2c152501027399a19364b1675003cc5/source/background.c#L1716

(and the same calculation in thermodynamics.c). I am not sure I understand this line, if the definition with respect to rho_tot is not a bug....

They also print out Omega0_lambda and Omega0_scf, properly using the rho_crit normalization.

**Comment by rainwoodman** [comment]
At this point I am convinced it is a bug on their side. Let's use rho_crit till they get back to us.

**Comment by nickhand** [comment]
Yeah I agree. I shall merge this now
